### PR TITLE
Name of receiver disappears #3204

### LINKED ISSUE #3204: Name of receiver disappears
The issue happens when there is a question about other teams AND one of other team names contains some special characters then we cannot see the name of receiver.

screenshot_3
In this case, the team name of student3 and student4 is __Team `~!@#$^&*()_-+={}[]\;:'"<>,.?/__. I'm sure that some of these characters cause the error. As far as I know, there are at least the character ' and \\.

It is not restricted in the question about other teams but all other questions as well.

screenshot_8
screenshot_9
screenshot_10
screenshot_11

And here is the question about other team.
screenshot_12

### PR DESCRIPTION
**Pull request by bamboo3250**
Fixes #3204 

screenshot_13

In feedbackSubmissionsEdit.js, the value of __curSelectedOption__ and __selectedOption__ may contain character ', which will make javascript confused.

for example: curSelectedOption = __team 1 '__
then the part __"option[value='"+curSelectedOption+"']"__ becomes __"option[value='team 1 '']"__

### CODE REVIEW

File: src/main/webapp/js/feedbackSubmissionsEdit.js [hunk 1]
```diff
@@ -416,6 +416,26 @@ function validateConstSumQuestions(){
...
+    return string.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1");
+}
+
+/**
+ * Replaces all occurences of regex by replace in str
+ */
+function replaceAll(regex, replace, str) {
+	return str.replace(new RegExp(escapeRegExp(regex), 'g'), replace);
+}
+
```
Comment (damithc) [inline src/main/webapp/js/feedbackSubmissionsEdit.js]: I would like to see unit tests for the above two functions. (We use QUnit to javascript unit tests)

**Comment by LowWeiLin** [comment]
The sanitization should not be done client-side. Is it possible to do it in the jsp/java files instead? Also, if the option is not a team, is the sanitization done as well(more than once)?

**Comment by bamboo3250** [comment]
@damithc I removed those 2 functions and made use of sanitizeForJs function.
@LowWeiLin __" Also, if the option is not a team, is the sanitization done as well(more than once)?"__
in that case, it should be either student or instructor and we use email address as the key for them. Therefore, there will be no difference between before and after applying __sanitizeForJs__ because characters such as ' and \ cannot be in the email address.

**Comment by bamboo3250** [comment]
Just to extend my answer, String variables are usually stored as sanitized string (for HTML). Here I use sanitization for Javascript and it still works for sanitized strings. Strings such as _&#39;_ or _'_ make no difference.

File: src/main/webapp/js/feedbackSubmissionsEdit.js [hunk 2]
```diff
@@ -447,7 +447,7 @@ function formatRecipientLists(){
             find("option[value='"+lastSelectedOption+"']").show();
         }
         if (curSelectedOption != "") {
-            $("select[name|="+FEEDBACK_RESPONSE_RECIPIENT+"-"+questionNumber+"]").not(this).
+        	$("select[name|="+FEEDBACK_RESPONSE_RECIPIENT+"-"+questionNumber+"]").not(this).
```
Comment (LowWeiLin) [inline src/main/webapp/js/feedbackSubmissionsEdit.js]: Is this change related? If not, please undo the changes.

Comment (bamboo3250) [inline src/main/webapp/js/feedbackSubmissionsEdit.js]: I accidently replaced 4 spaces by a tab. I will change it back when adding test case for this issue.

**Comment by bamboo3250** [comment]
@LowWeiLin I think we cannot fix this issue just by sanitizing from outside of Javascript. Here is the reason why:

screenshot_1
In __feedbackSubmissionEdit.js__, if __curSelectedOption__'s value is __Team>' 1__, Javascript will fail at the line 452 because the find() function had __option[value='Team>' 1']__ as parameter. (there are 3 quotes).
So I sanitize it (in .java file) then __curSelectedOption__ now is __Team>\' 1__
screenshot_2
But then the find() function I mentioned just now will only see it as __Team>' 1__ (without the slash \) due to the escape character. As a result, it cannot find the option with the value __Team>\' 1__ in the HTML file.

So my point is
If __curSelectedOption__ is __Team>' 1__, Javascript will fail
On the other hand, if __curSelectedOption__ is __Team>\' 1__, javascript will not fail but cannot find what we want. 
Even if __curSelectedOption__ is __Team>\\\' 1__, there is no change.

I also tried to use sanitization for HTML but it doesnt work. So I think we must have a sanitize function inside javascript like I did before.

**Comment by LowWeiLin** [comment]
Okay sure. Can do it with javascript and test with the cases you mentioned.

**Comment by bamboo3250** [comment]
Fixed main bug:
1. Added function __sanitizeForJs(string)__ in __common.js__
2. Fixed #3204 by using the function __sanitizeForJs(string)__ in __feedbackSubmissionsEdit.js__
3. Added test cases for the function __sanitizeForJs(string)__ in __CommonJsTest.js__

Added test case for the bug:
1. Changed team name in test data from __Team> 1__ to __Team >'"< 1__ in __StudentFeedbackSubmitPageUiTest.json__
2. Changed output files of the test case including:
        - __studentFeedbackSubmitPageFullyFilled.html__
        - __studentFeedbackSubmitPageModified.html__
        - __studentFeedbackSubmitPageOpen.html__
        - __studentFeedbackSubmitPagePartiallyFilled.html__
        - __unregisteredStudentFeedbackSubmitPageOpen.html__
        - __unregisteredStudentFeedbackSubmitPagePartiallyFilled.html__
3. Changed team name in the test driver __StudentFeedbackSubmitPageUiTest.java__
4. Here I found out that Selenium using XPath also got problem when handling single quote and double quote character so I added clearStringForXPath(String text) in __Sanitizer.java__
5. Used the function clearStringForXPath() in __FeedbackSubmitPage.java__ to pass the test
6. Added test case for clearStringForXPath(String text) in __SanitizerTest.java__
7. Used sanitizeForHtml in __HtmlHelper.java__ because I found out the data is not sanitized, possibly for injection and it cannot pass the test without doing this.
8. Added sanitizedForHtml in __feedbackSubmissionEdit.jsp__ as well because the data there is also not sanitized and it also cannot pass the test.
9. Dev green

File: src/main/java/teammates/common/util/Sanitizer.java [hunk 3]
```diff
@@ -175,4 +175,25 @@ public static String sanitizeForCsv(String str){
     private static String trimIfNotNull(String string) {
         return ((string == null) ? null : string.trim());
     }
+    
+    public static String clearStringForXPath(String text){
```
Comment (LowWeiLin) [inline src/main/java/teammates/common/util/Sanitizer.java]: Could you add comments on what this function does?

File: src/main/java/teammates/common/util/Sanitizer.java [hunk 4]
```diff
@@ -175,4 +175,25 @@ public static String sanitizeForCsv(String str){
...
+        String result = "";
+        int startPos = 0;
+        for(int i=0;i<text.length();i++){
+            while((i<text.length()) && (text.charAt(i)!='\'')) i++;
```
Comment (LowWeiLin) [inline src/main/java/teammates/common/util/Sanitizer.java]: Use curly braces for loops, according to the coding style.

**Comment by LowWeiLin** [comment]
If thee is a problem using the xpath, why can't we use ids to reference the web element instead?

**Comment by bamboo3250** [comment]
screenshot_1

These are radio buttons in HTML. We can see that there is no Id for these options. Therefore, the test unit has to use their values as ids.

screenshot_2

That's why the problem will raise when the team name contains those special character. Adding ids to these options is feasible but requires a lot of changes. So I decided to fix the XPath's problem instead.
